src/engine.rs

```diff
@@ -6,73 +6,133 @@ use crate::interpreter::*;
 use crate::lexer::*;
 use crate::parser::*;
 use crate::scheduler::*;
+use crate::utils::gather_functions;
 use crate::value::*;
 
 use anyhow::Result;
 
 #[derive(Clone)]
 pub struct Engine {
     modules: Vec<Ref<Module>>,
-    input: Value,
-    data: Value,
-}
-
-impl Default for Engine {
-    fn default() -> Self {
-        Self::new()
-    }
+    interpreter: Interpreter,
+    prepared: bool,
 }
 
 impl Engine {
-    pub fn new() -> Self {
-        Self {
+    pub fn new() -> Result<Self> {
```

Comment:
In future we can change this so that all our `new` functions return T instead of Result<T>; `new` functions are expected to always succeed.